### PR TITLE
Add a year long global simulation to run 1x per week

### PR DESCRIPTION
**Pull request by kmdeck**
## Purpose 
Add a year long global simulation to run 1x per week

Note that to minimize NaNs, I have decreased the resolution at the surface from to 50cm (in reality, we want something like 2.5 cm), and increased ac_canopy by 10x (in reality, ~10^3). Once we resolve the NaNs in this simulation through future PRs, we can decrease both of these.

Example output after 7 days:
soil water content
slw 604800 0
gpp
gpp 604800 0

## Content
Add a year long global simulation to run 1x per week [`pipeline.yml`, `experiments/long_runs/land.jl`, diagnostics code]

Use an updated set of soil parameters with the correct lat/lon and higher resolution [changes to `richards_runoff.jl`, `Artifacts.toml`, and soil params file name in `land.jl`]

Fix bug in bucket model which did not update drivers in callback [`global_bucket_function.jl`]

Update `NEWS.md`

## To-do
This PR:
- restore buildkite Manifest
- review

Future PRs - summarizing Alexis' comment below and adding some - 
1. Figure out NaNs - (a) if we make ac_canopy larger, fewer NaNs. (b) if we make the thermal timescale for phase change longer, we get fewer NaNs (c) if we make the timestep smaller, we get fewer NaNs -> issue with timestepping? better implicit solver? One other question: why NaNs, and not domain errors? 
2. Improve/debug diagnostics - more variables, perhaps we want monthly averages
3. Make better plots, maybe animations
4. Use mask in diagnostics so we only get continents
5. Use units that are reasonable for the community

Even later: in ClimaAnalysis, have a function to generate a local dashboard where we can use a slider to change time, a menu to change variables, maybe a play button for animations

----
- [ ] I have read and checked the items on the review checklist.

### CODE REVIEW

File: .buildkite/longruns_gpu/pipeline.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
+agents:
+  queue: clima
+  slurm_mem: 8G
+  modules: julia/1.10.0 cuda/julia-pref openmpi/4.1.5-mpitrampoline nsight-systems/2024.2.1
```
Comment (Sbozzolo) [inline .buildkite/longruns_gpu/pipeline.yml:4]:
```suggestion
  modules: common
```

File: .buildkite/longruns_gpu/pipeline.yml [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+env:
+  JULIA_MPI_HAS_CUDA: "true"
+  JULIA_NVTX_CALLBACKS: gc
+  JULIA_MAX_NUM_PRECOMPILE_FILES: 100
+  OPENBLAS_NUM_THREADS: 1
+  OMPI_MCA_opal_warn_on_missing_libcuda: 0
+  SLURM_KILL_BAD_EXIT: 1
+  SLURM_GRES_FLAGS: "allow-task-sharing"
+  CONFIG_PATH: "config/longrun_configs"
```
Comment (Sbozzolo) [inline .buildkite/longruns_gpu/pipeline.yml:14]:
```suggestion
env:
  JULIA_NVTX_CALLBACKS: gc
  JULIA_MAX_NUM_PRECOMPILE_FILES: 100
  OPENBLAS_NUM_THREADS: 1
  SLURM_KILL_BAD_EXIT: 1
  SLURM_GRES_FLAGS: "allow-task-sharing"
  CONFIG_PATH: "config/longrun_configs"
```

File: .buildkite/longruns_gpu/pipeline.yml [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      JULIA_MAX_NUM_PRECOMPILE_FILES: 50
+
+  - wait
+
+  - group: "Global Land Models"
+    steps:
+
+      - label: ":seedling: Soil-Canopy"
+        command:
+          - srun julia --color=yes --project=.buildkite experiments/long_runs/land.jl
```
Comment (Sbozzolo) [inline .buildkite/longruns_gpu/pipeline.yml:42]:
```suggestion
          - julia --color=yes --project=.buildkite experiments/long_runs/land.jl
```

File: .buildkite/longruns_gpu/pipeline.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  slurm_mem: 8G
+  modules: common
+
+env:
+  JULIA_NVTX_CALLBACKS: gc
+  JULIA_MAX_NUM_PRECOMPILE_FILES: 100
+  OPENBLAS_NUM_THREADS: 1
+  SLURM_KILL_BAD_EXIT: 1
+  SLURM_GRES_FLAGS: "allow-task-sharing"
+  CONFIG_PATH: "config/longrun_configs"
```
Comment (Sbozzolo) [inline .buildkite/longruns_gpu/pipeline.yml:12]: This is also not needed, it's used only by coupler/atmo

File: experiments/standalone/Bucket/global_bucket_function.jl [hunk 5]
```diff
@@ -177,11 +177,15 @@ diagnostic_handler =
 
 diag_cb = ClimaDiagnostics.DiagnosticsCallback(diagnostic_handler)
 
+updateat = collect(t0:(Δt * 3):tf);
```
Comment (kmdeck) [inline experiments/standalone/Bucket/global_bucket_function.jl:180]: we need to update the forcing in addition to the diagnostics cb

**Comment by Sbozzolo** [comment]
I scheduled this run to on Mondays starting at 23.35

(It would be good to name Buildkite pipelines without spaces)

File: experiments/long_runs/land.jl [hunk 6]
```diff
@@ -0,0 +1,660 @@
...
+    updatefunc = ClimaLand.make_update_drivers(atmos, radiation)
+
+    # ClimaDiagnostics
+
+    space = domain.space.subsurface
+
+    nc_writer = ClimaDiagnostics.Writers.NetCDFWriter(space, outdir)
+
+    diags =
+        ClimaLand.CLD.default_diagnostics(model, t0; output_writer = nc_writer)
```
Comment (kmdeck) [inline experiments/long_runs/land.jl:609]: model = land

File: experiments/long_runs/land.jl [hunk 7]
```diff
@@ -0,0 +1,660 @@
...
+    )
+
+    updateat = Array(t0:(3600 * 3):tf)
+    updatefunc = ClimaLand.make_update_drivers(atmos, radiation)
+
+    # ClimaDiagnostics
+
+    space = domain.space.subsurface
+
+    nc_writer = ClimaDiagnostics.Writers.NetCDFWriter(space, outdir)
```
Comment (kmdeck) [inline experiments/long_runs/land.jl:606]: already defined above

Comment (AlexisRenchon) [reply experiments/long_runs/land.jl:606]: yes, subsurface_space

File: experiments/long_runs/land.jl [hunk 8]
```diff
@@ -0,0 +1,660 @@
...
+
+    diags =
+        ClimaLand.CLD.default_diagnostics(model, t0; output_writer = nc_writer)
+
+    diagnostic_handler =
+        ClimaDiagnostics.DiagnosticsHandler(diags, Y, p, t0; dt = Δt)
```
Comment (AlexisRenchon) [inline experiments/long_runs/land.jl:612]: dt = dt

File: experiments/standalone/Bucket/global_bucket_function.jl [hunk 9]
```diff
@@ -177,11 +177,15 @@ diagnostic_handler =
...
+updatefunc = ClimaLand.make_update_drivers(bucket_atmos, bucket_rad)
+driver_cb = ClimaLand.DriverUpdateCallback(updateat, updatefunc)
+
 sol = ClimaComms.@time ClimaComms.device() SciMLBase.solve(
     prob,
     ode_algo;
     dt = Δt,
```
Comment (AlexisRenchon) [inline experiments/standalone/Bucket/global_bucket_function.jl:187]: dt = dt

File: experiments/long_runs/land.jl (hunk 6 above)
Comment (AlexisRenchon) [reply experiments/long_runs/land.jl:609]: diags =
        ClimaLand.CLD.default_diagnostics(model, t0; output_writer = nc_writer, output_vars = "short")

see the changes in src/diagnostics/default_diagnostics.jl in PR #682 (for now, copy the PR files to this PR?)

**Comment by AlexisRenchon** [comment]
I can fix the diagnostics on Friday the latest! 
We can coordinate and I'd directly push my commit to this PR

File: experiments/long_runs/land.jl [hunk 10]
```diff
@@ -0,0 +1,660 @@
...
+        depth = depth,
+        nelements = nelements,
+        npolynomial = 1,
+        dz_tuple = FT.((10.0, 0.1)),
+    )
+    surface_space = domain.space.surface
+    subsurface_space = domain.space.subsurface
+
+    ref_time = DateTime(2021)
+    t_start = 0.0
```
Comment (AlexisRenchon) [inline experiments/long_runs/land.jl:76]: Is this the same as t0? (which is an argument to the `setup_prob` function)

Comment (kmdeck) [reply experiments/long_runs/land.jl:76]: yes, good catch!! let's use the same convetnion. i think they are the same

File: experiments/long_runs/land.jl [hunk 11]
```diff
@@ -0,0 +1,658 @@
...
+
+regridder_type = :InterpolationsRegridder
+context = ClimaComms.context()
+device = ClimaComms.device()
+device_suffix = device isa ClimaComms.CPUSingleThreaded ? "cpu" : "gpu"
+root_path = "land_longrun_$(device_suffix)"
+diagnostics_outdir = joinpath(root_path, "global_diagnostics")
+outdir =
+    ClimaUtilities.OutputPathGenerator.generate_output_path(diagnostics_outdir)
+!ispath(outdir) && mkpath(outdir)
```
Comment (AlexisRenchon) [inline experiments/long_runs/land.jl:58]: What does this do? it is necessary?

Comment (kmdeck) [reply experiments/long_runs/land.jl:58]: oh, perhaps not. Does `ClimaUtilities.OutputPathGenerator.generate_output_path`? make the directory/path to the diagnostics?  If so, we dont need the second line.

This line is a compact but hard to read way of making the path:
```
!ispath(outdir) && mkpath(outdir)
```
1. !ispath(outdir) = true if the path does not exist
2. then evaluates mkpath(outdir)
If the first statement is false (it exists), it does not evaluated the 2nd clause, I think

**Comment by AlexisRenchon** [comment]
Ok I made some changes and pushed my commits, GPP diagnostics is written, and plots are produced, it runs well on the cluster. 

However, GPP becomes NaN progressively? (not sure if this is a simulation bug or a diagnostic writing bug?) 

I think next step for this PR (or a next PR) is:

- [ ] add other diagnostic variables (need to fix things a bit in src/diagnostics, e.g., I forgot a lot of ! at the end of compute function name)
- [ ] make better plots, maybe animations?
- [ ] fix the bugs (NaNs for examples) and make the simulation 1 year long

And for later, maybe in ClimaAnalysis, have a function to generate a local dashboard where we can use a slider to change time, a menu to change variables, maybe a play button for animations?

gpp 46800 0

File: experiments/long_runs/land.jl [hunk 12]
```diff
@@ -0,0 +1,662 @@
...
+    # Define timestepper and ODE algorithm
+    stepper = CTS.ARS343()
+    ode_algo = CTS.IMEXAlgorithm(
+        stepper,
+        CTS.NewtonsMethod(
+            max_iters = 1,
+            update_j = CTS.UpdateEvery(CTS.NewTimeStep),
+        ),
+    )
+    SciMLBase.solve(prob, ode_algo; dt = Δt, callback = cb)
```
Comment (kmdeck) [inline experiments/long_runs/land.jl:642]: adaptive=false?

File: .buildkite/longruns_gpu/pipeline.yml [hunk 13]
```diff
@@ -0,0 +1,39 @@
...
+  slurm_mem: 8G
+  modules: common
+
+timeout_in_minutes: 1440
+
+steps:
+  - label: "init :GPU:"
+    key: "init_gpu_env"
+    command:
+      - echo "--- Instantiate"
```
Comment (kmdeck) [inline .buildkite/longruns_gpu/pipeline.yml:12]: update this

File: NEWS.md [hunk 14]
```diff
@@ -3,15 +3,16 @@ ClimaLand.jl Release Notes
 
 main
 --------
+<<<<<<< HEAD
```
Comment (kmdeck) [inline NEWS.md:6]: remove

File: .buildkite/Manifest.toml [hunk 15]
```diff
@@ -1,6 +1,6 @@
 # This file is machine-generated - editing it directly is not advised
 
-julia_version = "1.10.4"
+julia_version = "1.10.3"
```
Comment (kmdeck) [inline .buildkite/Manifest.toml:3]: Ill restore this

File: Artifacts.toml [hunk 16]
```diff
@@ -1,9 +1,12 @@
 [era5_land_forcing_data2021]
 git-tree-sha1 = "ec424296df6b60cfe273ac8f981701fbbed0bd8a"
 
+["processed_topographic_index 2.5 deg"]
```
Comment (kmdeck) [inline Artifacts.toml:4]: will remove this one

**Review by AlexisRenchon — APPROVED** [review]

**Comment by AlexisRenchon** [comment]
Just a thought that eventually the weekly CI run of this could push a web dashboard to a server (the existing Microsoft Azure VM or another). 

I think that would be very cool. We will always be able to interact with global outputs of latest run from the web, convenient for discussions etc. I might work on this with the viz students this summer...

**Pull request merged**